### ISSUE TITLE
closured function fails with RecursionError

### ISSUE THREAD

**Issue by mmckerns** [uqfoundation/dill#211]
Originally reported by @jetheron in #129.

```
import dill

def foo():
    def _foo():
        if not hasattr(_foo, '_counter'):
            _foo._counter = 0
        _foo._counter += 1
        return _foo._counter
    return _foo

f = foo()
f()
f()
dill.dumps(f)
```

**Comment by mmckerns** [comment]
@matsjoyce: I think this is one issue we run into all the time, and here's a solution: 
https://medium.com/the-infinite-machine/python-function-serialisation-with-yccloudpickle-b2ff6b2ad5da
https://medium.com/@emlynoregan/automatically-serialising-recursive-inner-functions-in-python-using-the-y-combinator-fc5d37e50b29
https://github.com/emlynoregan/yccloudpickle

**Comment by mmckerns** [comment]
fixed by #443

**Comment by mmckerns** [comment]
Hi @tvalentyn.  Yes -- but for a number of reasons (that I won't go into here) it's about 4 months late.  However (and especially now that #469 is closed) it is literally the next thing I'm planning to do... and hopefully that means it will be out by this afternoon.
